lib/fable/events.ex

```diff
@@ -105,43 +205,73 @@ defmodule Fable.Events do
     |> Enum.map(&Fable.Event.parse_data(config.repo, &1))
   end
 
-  def emit!(config, schema, event, opts \\ []) do
-    case emit(config, schema, event, opts) do
-      {:ok, schema} ->
-        schema
+  @doc false
+  @spec emit(Fable.Config.t(), aggregate_source, event_or_events, Keyword.t()) :: transation_fun
+  def unconditionally_emit(config, schema_or_queryable, event_or_events, opts \\ []) do
+    emit(config, schema_or_queryable, fn _, _, _ -> event_or_events end, opts)
+  end
+
+  @doc false
+  @spec emit(Fable.Config.t(), aggregate_source, emit_callback, Ecto.Multi.changes(), Keyword.t()) ::
+          transation_fun
 
-      error ->
-        raise error
+  def emit(config, schema_or_queryable, fun, changes \\ %{}, opts)
+
+  def emit(%{repo: repo} = config, %Ecto.Query{} = queryable, fun, changes, opts)
+      when is_function(fun, 3) do
+    fn ->
+      schema = lock(queryable, repo) || raise "No aggregate found"
+      check_schema_fields(schema)
+      rollback_on_error(repo, handle_events(config, schema, fun.(schema, repo, changes), opts))
     end
   end
 
-  def unconditionally_emit(%{repo: repo} = config, schema_or_queryable, event, opts \\ []) do
-    # Unlike emit/4, we want to emit this event regardless of whether we've been
-    # passed an out of date schema
-    repo.serial(schema_or_queryable, fn schema ->
-      do_emit(config, schema, event, opts)
-    end)
+  def emit(
+        %{repo: repo} = config,
+        %{__meta__: %Ecto.Schema.Metadata{}} = schema,
+        fun,
+        changes,
+        opts
+      )
+      when is_function(fun, 3) do
+    check_schema_fields(schema)
+
+    fn ->
+      schema = lock(schema, repo) || schema
+      rollback_on_error(repo, handle_events(config, schema, fun.(schema, repo, changes), opts))
```

Comment:
as a minor style thing, I'd probably prefer to yank out the `fun.()` call and the `handle_events` call here into named variables.